File: ts/src/meshtrade/type/v1/time_of_day.test.ts
import {
  newTimeOfDay,
  newTimeOfDayFromJsDate,
  newTimeOfDayFromMillis,
  timeOfDayToMillis,
  timeOfDayToJsDateWithDate,
  isValid,
  isMidnight,
  timeOfDayToString,
  totalSeconds,
} from "./time_of_day";
import { TimeOfDay, TimeOfDaySchema } from "./time_of_day_pb";
import { newDate } from "./date";
import { DateSchema } from "./date_pb";
import { create } from "@bufbuild/protobuf";

describe("newTimeOfDay", () => {
  // Valid test cases as individual tests
  test("valid midnight", () => {
    const result = newTimeOfDay(0, 0, 0, 0);
    expect(result.hours).toBe(0);
    expect(result.minutes).toBe(0);
    expect(result.seconds).toBe(0);
    expect(result.nanos).toBe(0);
  });

  test("valid noon", () => {
    const result = newTimeOfDay(12, 0, 0, 0);
    expect(result.hours).toBe(12);
    expect(result.minutes).toBe(0);
    expect(result.seconds).toBe(0);
    expect(result.nanos).toBe(0);
  });

  test("valid end of day", () => {
    const result = newTimeOfDay(23, 59, 59, 999999999);
    expect(result.hours).toBe(23);
    expect(result.minutes).toBe(59);
    expect(result.seconds).toBe(59);
    expect(result.nanos).toBe(999999999);
  });

  // Invalid test cases as test.each
  const invalidTestCases: Array<
    [string, number, number, number, number, string?]
  > = [
      ["invalid hours too low", -1, 0, 0, 0, "Hours must be between 0 and 23"],
      ["invalid hours too high", 24, 0, 0, 0, "Hours must be between 0 and 23"],
      [
        "invalid minutes too low",
        12,
        -1,
        0,
        0,
        "Minutes must be between 0 and 59",
      ],
      [
        "invalid minutes too high",
        12,
        60,
        0,
        0,
        "Minutes must be between 0 and 59",
      ],
      [
        "invalid seconds too low",
        12,
        30,
        -1,
        0,
        "Seconds must be between 0 and 59",
      ],
      [
        "invalid seconds too high",
        12,
        30,
        60,
        0,
        "Seconds must be between 0 and 59",
      ],
      [
        "invalid nanos too low",
        12,
        30,
        45,
        -1,
        "Nanos must be between 0 and 999,999,999",
      ],
      [
        "invalid nanos too high",
        12,
        30,
        45,
        1000000000,
        "Nanos must be between 0 and 999,999,999",
      ],
    ];

  test.each(invalidTestCases)(
    "%s",
    (_, hours, minutes, seconds, nanos, errorMessage) => {
      expect(() => newTimeOfDay(hours, minutes, seconds, nanos)).toThrow();
      if (errorMessage) {
        expect(() => newTimeOfDay(hours, minutes, seconds, nanos)).toThrow(
          errorMessage
        );
      }
    }
  );
});

describe("newTimeOfDayFromJsDate", () => {
  test("midnight", () => {
    const result = newTimeOfDayFromJsDate(new Date(2023, 11, 25, 0, 0, 0, 0));
    expect(result.hours).toBe(0);
    expect(result.minutes).toBe(0);
    expect(result.seconds).toBe(0);
    expect(result.nanos).toBe(0);
  });

  test("noon", () => {
    const result = newTimeOfDayFromJsDate(new Date(2023, 11, 25, 12, 0, 0, 0));
    expect(result.hours).toBe(12);
    expect(result.minutes).toBe(0);
    expect(result.seconds).toBe(0);
    expect(result.nanos).toBe(0);
  });

  test("complex time", () => {
    const result = newTimeOfDayFromJsDate(
      new Date(2023, 11, 25, 15, 30, 45, 123)
    );
    expect(result.hours).toBe(15);
    expect(result.minutes).toBe(30);
    expect(result.seconds).toBe(45);
    expect(result.nanos).toBe(123000000);
  });

  test("end of day", () => {
    const result = newTimeOfDayFromJsDate(
      new Date(2023, 11, 25, 23, 59, 59, 999)
    );
    expect(result.hours).toBe(23);
    expect(result.minutes).toBe(59);
    expect(result.seconds).toBe(59);
    expect(result.nanos).toBe(999000000);
  });
});

describe("newTimeOfDayFromMillis", () => {
  // Valid test cases as individual tests
  test("midnight", () => {
    const result = newTimeOfDayFromMillis(0);
    expect(result.hours).toBe(0);
    expect(result.minutes).toBe(0);
    expect(result.seconds).toBe(0);
    expect(result.nanos).toBe(0);
  });

  test("one hour", () => {
    const result = newTimeOfDayFromMillis(3600000);
    expect(result.hours).toBe(1);
    expect(result.minutes).toBe(0);
    expect(result.seconds).toBe(0);
    expect(result.nanos).toBe(0);
  });

  test("complex time", () => {
    const result = newTimeOfDayFromMillis(
      2 * 3600000 + 30 * 60000 + 45 * 1000 + 123
    );
    expect(result.hours).toBe(2);
    expect(result.minutes).toBe(30);
    expect(result.seconds).toBe(45);
    expect(result.nanos).toBe(123000000);
  });

  test("almost 24 hours", () => {
    const result = newTimeOfDayFromMillis(24 * 3600000 - 1);
    expect(result.hours).toBe(23);
    expect(result.minutes).toBe(59);
    expect(result.seconds).toBe(59);
    expect(result.nanos).toBe(999000000);
  });

  // Invalid test cases as test.each
  const invalidTestCases: Array<[string, number, string]> = [
    ["negative milliseconds", -1000, "Milliseconds cannot be negative"],
    [
      "exactly 24 hours",
      24 * 3600000,
      "Milliseconds cannot be 24 hours or more",
    ],
    [
      "more than 24 hours",
      25 * 3600000,
      "Milliseconds cannot be 24 hours or more",
    ],
  ];

  test.each(invalidTestCases)("%s", (_, millis, errorMessage) => {
    expect(() => newTimeOfDayFromMillis(millis)).toThrow();
    expect(() => newTimeOfDayFromMillis(millis)).toThrow(errorMessage);
  });
});

describe("timeOfDayToMillis", () => {
  test("undefined time", () => {
    // Cast to any to test undefined case
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const result = timeOfDayToMillis(undefined as any);
    expect(result).toBe(0);
  });

  test("midnight", () => {
    const result = timeOfDayToMillis(newTimeOfDay(0, 0, 0, 0));
    expect(result).toBe(0);
  });

  test("noon", () => {
    const result = timeOfDayToMillis(newTimeOfDay(12, 0, 0, 0));
    expect(result).toBe(12 * 3600000);
  });

  test("one minute", () => {
    const result = timeOfDayToMillis(newTimeOfDay(0, 1, 0, 0));
    expect(result).toBe(60000);
  });

  test("one second", () => {
    const result = timeOfDayToMillis(newTimeOfDay(0, 0, 1, 0));
    expect(result).toBe(1000);
  });

  test("complex time", () => {
    const result = timeOfDayToMillis(newTimeOfDay(2, 30, 45, 123000000));
    expect(result).toBe(2 * 3600000 + 30 * 60000 + 45 * 1000 + 123);
  });

  test("end of day", () => {
    const result = timeOfDayToMillis(newTimeOfDay(23, 59, 59, 999000000));
    expect(result).toBe(23 * 3600000 + 59 * 60000 + 59 * 1000 + 999);
  });
});

describe("timeOfDayToJsDateWithDate", () => {
  const date = newDate(2023, 12, 25);

  // Valid test cases as individual tests
  test("valid datetime", () => {
    const timeOfDay = newTimeOfDay(15, 30, 45, 123000000);
    const result = timeOfDayToJsDateWithDate(timeOfDay, date);
    expect(result).toBeInstanceOf(Date);
    expect(result.getFullYear()).toBe(date.year);
    expect(result.getMonth()).toBe(date.month - 1); // JS months are 0-indexed
    expect(result.getDay()).toBe(date.day);
    expect(result.getHours()).toBe(timeOfDay.hours);
    expect(result.getMinutes()).toBe(timeOfDay.minutes);
    expect(result.getSeconds()).toBe(timeOfDay.seconds);
  });

  test("midnight on date", () => {
    const timeOfDay = newTimeOfDay(0, 0, 0, 0);
    const result = timeOfDayToJsDateWithDate(timeOfDay, date);
    expect(result).toBeInstanceOf(Date);
    expect(result.getFullYear()).toBe(date.year);
    expect(result.getMonth()).toBe(date.month - 1); // JS months are 0-indexed
    expect(result.getDate()).toBe(date.day);
    expect(result.getHours()).toBe(timeOfDay.hours);
    expect(result.getMinutes()).toBe(timeOfDay.minutes);
    expect(result.getSeconds()).toBe(timeOfDay.seconds);
  });

  // Invalid test cases as test.each
  const invalidTestCases: Array<
    [string, TimeOfDay | undefined, typeof date | undefined, string]
  > = [
      [
        "undefined time",
        undefined,
        date,
        "TimeOfDay object is null or undefined",
      ],
      [
        "undefined date",
        newTimeOfDay(12, 0, 0, 0),
        undefined,
        "Date object is null or undefined",
      ],
      [
        "incomplete date",
        newTimeOfDay(12, 0, 0, 0),
        (() => {
          const incompleteDate = create(DateSchema)
          incompleteDate.year = 2023;
          incompleteDate.month = 1;
          // Day is not set, so it remains 0, making the date incomplete
          return incompleteDate;
        })(),
        "Date must be complete",
      ],
    ];

  test.each(invalidTestCases)("%s", (_, timeOfDay, protoDate, errorMessage) => {
    expect(() => timeOfDayToJsDateWithDate(timeOfDay!, protoDate!)).toThrow();
    expect(() => timeOfDayToJsDateWithDate(timeOfDay!, protoDate!)).toThrow(
      errorMessage
    );
  });
});

describe("isValid", () => {
  test("undefined time", () => {
    expect(isValid(undefined)).toBe(false);
  });

  test("valid midnight", () => {
    expect(isValid(newTimeOfDay(0, 0, 0, 0))).toBe(true);
  });

  test("valid noon", () => {
    expect(isValid(newTimeOfDay(12, 0, 0, 0))).toBe(true);
  });

  test("valid end of day", () => {
    expect(isValid(newTimeOfDay(23, 59, 59, 999999999))).toBe(true);
  });

  test("invalid hours", () => {
    const timeOfDay = create(TimeOfDaySchema, {
      hours: 12,
      minutes: 0,
      seconds: 0,
      nanos: 0,
    })
    expect(isValid(timeOfDay)).toBe(false);
  });

  test("invalid minutes", () => {
    const timeOfDay = create(TimeOfDaySchema, {
      hours: 12,
      minutes: 60,
      seconds: 0,
      nanos: 0,
    })
    expect(isValid(timeOfDay)).toBe(false);
  });

  test("invalid seconds", () => {
    const timeOfDay = create(TimeOfDaySchema, {
      hours: 12,
      minutes: 30,
      seconds: 60,
      nanos: 0,
    })
    expect(isValid(timeOfDay)).toBe(false);
  });

  test("invalid nanos", () => {
    const timeOfDay = create(TimeOfDaySchema, {
      hours: 12,
      minutes: 30,
      seconds: 45,
      nanos: 1000000000,
    })
    expect(isValid(timeOfDay)).toBe(false);
  });
});

describe("isMidnight", () => {
  test("undefined time", () => {
    expect(isMidnight(undefined)).toBe(false);
  });

  test("midnight", () => {
    expect(isMidnight(newTimeOfDay(0, 0, 0, 0))).toBe(true);
  });

  test("not midnight - 1 hour", () => {
    expect(isMidnight(newTimeOfDay(1, 0, 0, 0))).toBe(false);
  });

  test("not midnight - 1 minute", () => {
    expect(isMidnight(newTimeOfDay(0, 1, 0, 0))).toBe(false);
  });

  test("not midnight - 1 second", () => {
    expect(isMidnight(newTimeOfDay(0, 0, 1, 0))).toBe(false);
  });

  test("not midnight - 1 nano", () => {
    expect(isMidnight(newTimeOfDay(0, 0, 0, 1))).toBe(false);
  });
});

describe("timeOfDayToString", () => {
  test("undefined time", () => {
    expect(timeOfDayToString(undefined)).toBe("<undefined>");
  });

  test("midnight", () => {
    expect(timeOfDayToString(newTimeOfDay(0, 0, 0, 0))).toBe("00:00:00");
  });

  test("noon", () => {
    expect(timeOfDayToString(newTimeOfDay(12, 0, 0, 0))).toBe("12:00:00");
  });

  test("complex time without nanos", () => {
    expect(timeOfDayToString(newTimeOfDay(15, 30, 45, 0))).toBe("15:30:45");
  });

  test("complex time with nanos", () => {
    expect(timeOfDayToString(newTimeOfDay(15, 30, 45, 123456789))).toBe(
      "15:30:45.123456789"
    );
  });

  test("single digits", () => {
    expect(timeOfDayToString(newTimeOfDay(1, 2, 3, 4))).toBe(
      "01:02:03.000000004"
    );
  });

  test("end of day", () => {
    expect(timeOfDayToString(newTimeOfDay(23, 59, 59, 999999999))).toBe(
      "23:59:59.999999999"
    );
  });
});

describe("totalSeconds", () => {
  test("undefined time", () => {
    const result = totalSeconds(undefined);
    expect(result).toBeCloseTo(0, 9);
  });

  test("midnight", () => {
    const result = totalSeconds(newTimeOfDay(0, 0, 0, 0));
    expect(result).toBeCloseTo(0, 9);
  });

  test("one hour", () => {
    const result = totalSeconds(newTimeOfDay(1, 0, 0, 0));
    expect(result).toBeCloseTo(3600, 9);
  });

  test("one minute", () => {
    const result = totalSeconds(newTimeOfDay(0, 1, 0, 0));
    expect(result).toBeCloseTo(60, 9);
  });

  test("one second", () => {
    const result = totalSeconds(newTimeOfDay(0, 0, 1, 0));
    expect(result).toBeCloseTo(1, 9);
  });

  test("half second", () => {
    const result = totalSeconds(newTimeOfDay(0, 0, 0, 500000000));
    expect(result).toBeCloseTo(0.5, 9);
  });

  test("complex time", () => {
    const result = totalSeconds(newTimeOfDay(2, 30, 45, 123456789));
    expect(result).toBeCloseTo(2 * 3600 + 30 * 60 + 45 + 0.123456789, 9);
  });

  test("end of day", () => {
    const result = totalSeconds(newTimeOfDay(23, 59, 59, 999999999));
    expect(result).toBeCloseTo(23 * 3600 + 59 * 60 + 59 + 0.999999999, 9);
  });
});

File: go/type/v1/time_of_day_test.go
package type_v1

import (
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestNewTimeOfDay(t *testing.T) {
	tests := []struct {
		name      string
		hours     int32
		minutes   int32
		seconds   int32
		nanos     int32
		wantErr   bool
		wantTime  *TimeOfDay
		errSubstr string
	}{
		{
			name:     "valid time midnight",
			hours:    0,
			minutes:  0,
			seconds:  0,
			nanos:    0,
			wantErr:  false,
			wantTime: &TimeOfDay{Hours: 0, Minutes: 0, Seconds: 0, Nanos: 0},
		},
		{
			name:     "valid time noon",
			hours:    12,
			minutes:  0,
			seconds:  0,
			nanos:    0,
			wantErr:  false,
			wantTime: &TimeOfDay{Hours: 12, Minutes: 0, Seconds: 0, Nanos: 0},
		},
		{
			name:     "valid time end of day",
			hours:    23,
			minutes:  59,
			seconds:  59,
			nanos:    999999999,
			wantErr:  false,
			wantTime: &TimeOfDay{Hours: 23, Minutes: 59, Seconds: 59, Nanos: 999999999},
		},
		{
			name:      "invalid hours too low",
			hours:     -1,
			minutes:   0,
			seconds:   0,
			nanos:     0,
			wantErr:   true,
			errSubstr: "hours must be between 0 and 23",
		},
		{
			name:      "invalid hours too high",
			hours:     24,
			minutes:   0,
			seconds:   0,
			nanos:     0,
			wantErr:   true,
			errSubstr: "hours must be between 0 and 23",
		},
		{
			name:      "invalid minutes too low",
			hours:     12,
			minutes:   -1,
			seconds:   0,
			nanos:     0,
			wantErr:   true,
			errSubstr: "minutes must be between 0 and 59",
		},
		{
			name:      "invalid minutes too high",
			hours:     12,
			minutes:   60,
			seconds:   0,
			nanos:     0,
			wantErr:   true,
			errSubstr: "minutes must be between 0 and 59",
		},
		{
			name:      "invalid seconds too low",
			hours:     12,
			minutes:   30,
			seconds:   -1,
			nanos:     0,
			wantErr:   true,
			errSubstr: "seconds must be between 0 and 59",
		},
		{
			name:      "invalid seconds too high",
			hours:     12,
			minutes:   30,
			seconds:   60,
			nanos:     0,
			wantErr:   true,
			errSubstr: "seconds must be between 0 and 59",
		},
		{
			name:      "invalid nanos too low",
			hours:     12,
			minutes:   30,
			seconds:   45,
			nanos:     -1,
			wantErr:   true,
			errSubstr: "nanos must be between 0 and 999,999,999",
		},
		{
			name:      "invalid nanos too high",
			hours:     12,
			minutes:   30,
			seconds:   45,
			nanos:     1000000000,
			wantErr:   true,
			errSubstr: "nanos must be between 0 and 999,999,999",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, err := NewTimeOfDay(tt.hours, tt.minutes, tt.seconds, tt.nanos)
			if tt.wantErr {
				require.Error(t, err)
				assert.Contains(t, err.Error(), tt.errSubstr)
				assert.Nil(t, got)
			} else {
				require.NoError(t, err)
				assert.Equal(t, tt.wantTime, got)
			}
		})
	}
}

func TestNewTimeOfDayFromDuration(t *testing.T) {
	tests := []struct {
		name      string
		duration  time.Duration
		wantErr   bool
		wantTime  *TimeOfDay
		errSubstr string
	}{
		{
			name:     "midnight",
			duration: 0,
			wantErr:  false,
			wantTime: &TimeOfDay{Hours: 0, Minutes: 0, Seconds: 0, Nanos: 0},
		},
		{
			name:     "one hour",
			duration: time.Hour,
			wantErr:  false,
			wantTime: &TimeOfDay{Hours: 1, Minutes: 0, Seconds: 0, Nanos: 0},
		},
		{
			name:     "complex time",
			duration: 2*time.Hour + 30*time.Minute + 45*time.Second + 123456789*time.Nanosecond,
			wantErr:  false,
			wantTime: &TimeOfDay{Hours: 2, Minutes: 30, Seconds: 45, Nanos: 123456789},
		},
		{
			name:     "almost 24 hours",
			duration: 23*time.Hour + 59*time.Minute + 59*time.Second + 999999999*time.Nanosecond,
			wantErr:  false,
			wantTime: &TimeOfDay{Hours: 23, Minutes: 59, Seconds: 59, Nanos: 999999999},
		},
		{
			name:      "negative duration",
			duration:  -time.Hour,
			wantErr:   true,
			errSubstr: "duration cannot be negative",
		},
		{
			name:      "exactly 24 hours",
			duration:  24 * time.Hour,
			wantErr:   true,
			errSubstr: "duration cannot be 24 hours or more",
		},
		{
			name:      "more than 24 hours",
			duration:  25 * time.Hour,
			wantErr:   true,
			errSubstr: "duration cannot be 24 hours or more",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, err := NewTimeOfDayFromDuration(tt.duration)
			if tt.wantErr {
				require.Error(t, err)
				assert.Contains(t, err.Error(), tt.errSubstr)
				assert.Nil(t, got)
			} else {
				require.NoError(t, err)
				assert.Equal(t, tt.wantTime, got)
			}
		})
	}
}

func TestTimeOfDay_ToDuration(t *testing.T) {
	tests := []struct {
		name      string
		timeOfDay *TimeOfDay
		want      time.Duration
	}{
		{
			name:      "nil time",
			timeOfDay: nil,
			want:      0,
		},
		{
			name:      "midnight",
			timeOfDay: &TimeOfDay{Hours: 0, Minutes: 0, Seconds: 0, Nanos: 0},
			want:      0,
		},
		{
			name:      "noon",
			timeOfDay: &TimeOfDay{Hours: 12, Minutes: 0, Seconds: 0, Nanos: 0},
			want:      12 * time.Hour,
		},
		{
			name:      "complex time",
			timeOfDay: &TimeOfDay{Hours: 2, Minutes: 30, Seconds: 45, Nanos: 123456789},
			want:      2*time.Hour + 30*time.Minute + 45*time.Second + 123456789*time.Nanosecond,
		},
		{
			name:      "end of day",
			timeOfDay: &TimeOfDay{Hours: 23, Minutes: 59, Seconds: 59, Nanos: 999999999},
			want:      23*time.Hour + 59*time.Minute + 59*time.Second + 999999999*time.Nanosecond,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := tt.timeOfDay.ToDuration()
			assert.Equal(t, tt.want, got)
		})
	}
}

func TestTimeOfDay_ToTimeWithDate(t *testing.T) {
	tests := []struct {
		name      string
		timeOfDay *TimeOfDay
		date      *Date
		wantErr   bool
		wantTime  time.Time
		errSubstr string
	}{
		{
			name:      "valid datetime",
			timeOfDay: &TimeOfDay{Hours: 15, Minutes: 30, Seconds: 45, Nanos: 123456789},
			date:      &Date{Year: 2023, Month: 12, Day: 25},
			wantErr:   false,
			wantTime:  time.Date(2023, 12, 25, 15, 30, 45, 123456789, time.UTC),
		},
		{
			name:      "midnight on date",
			timeOfDay: &TimeOfDay{Hours: 0, Minutes: 0, Seconds: 0, Nanos: 0},
			date:      &Date{Year: 2023, Month: 1, Day: 1},
			wantErr:   false,
			wantTime:  time.Date(2023, 1, 1, 0, 0, 0, 0, time.UTC),
		},
		{
			name:      "nil time of day",
			timeOfDay: nil,
			date:      &Date{Year: 2023, Month: 12, Day: 25},
			wantErr:   true,
			errSubstr: "time of day is nil",
		},
		{
			name:      "nil date",
			timeOfDay: &TimeOfDay{Hours: 12, Minutes: 0, Seconds: 0, Nanos: 0},
			date:      nil,
			wantErr:   true,
			errSubstr: "date must be complete",
		},
		{
			name:      "incomplete date",
			timeOfDay: &TimeOfDay{Hours: 12, Minutes: 0, Seconds: 0, Nanos: 0},
			date:      &Date{Year: 2023, Month: 0, Day: 0},
			wantErr:   true,
			errSubstr: "date must be complete",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, err := tt.timeOfDay.ToTimeWithDate(tt.date)
			if tt.wantErr {
				require.Error(t, err)
				assert.Contains(t, err.Error(), tt.errSubstr)
			} else {
				require.NoError(t, err)
				assert.Equal(t, tt.wantTime, got)
			}
		})
	}
}

func TestTimeOfDay_IsValid(t *testing.T) {
	tests := []struct {
		name      string
		timeOfDay *TimeOfDay
		want      bool
	}{
		{"nil time", nil, false},
		{"valid midnight", &TimeOfDay{Hours: 0, Minutes: 0, Seconds: 0, Nanos: 0}, true},
		{"valid noon", &TimeOfDay{Hours: 12, Minutes: 0, Seconds: 0, Nanos: 0}, true},
		{"valid end of day", &TimeOfDay{Hours: 23, Minutes: 59, Seconds: 59, Nanos: 999999999}, true},
		{"invalid hours", &TimeOfDay{Hours: 24, Minutes: 0, Seconds: 0, Nanos: 0}, false},
		{"invalid minutes", &TimeOfDay{Hours: 12, Minutes: 60, Seconds: 0, Nanos: 0}, false},
		{"invalid seconds", &TimeOfDay{Hours: 12, Minutes: 30, Seconds: 60, Nanos: 0}, false},
		{"invalid nanos", &TimeOfDay{Hours: 12, Minutes: 30, Seconds: 45, Nanos: 1000000000}, false},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := tt.timeOfDay.IsValid()
			assert.Equal(t, tt.want, got)
		})
	}
}

func TestTimeOfDay_TotalSeconds(t *testing.T) {
	tests := []struct {
		name      string
		timeOfDay *TimeOfDay
		want      float64
	}{
		{
			name:      "nil time",
			timeOfDay: nil,
			want:      0,
		},
		{
			name:      "midnight",
			timeOfDay: &TimeOfDay{Hours: 0, Minutes: 0, Seconds: 0, Nanos: 0},
			want:      0,
		},
		{
			name:      "one hour",
			timeOfDay: &TimeOfDay{Hours: 1, Minutes: 0, Seconds: 0, Nanos: 0},
			want:      3600,
		},
		{
			name:      "one minute",
			timeOfDay: &TimeOfDay{Hours: 0, Minutes: 1, Seconds: 0, Nanos: 0},
			want:      60,
		},
		{
			name:      "one second",
			timeOfDay: &TimeOfDay{Hours: 0, Minutes: 0, Seconds: 1, Nanos: 0},
			want:      1,
		},
		{
			name:      "half second",
			timeOfDay: &TimeOfDay{Hours: 0, Minutes: 0, Seconds: 0, Nanos: 500000000},
			want:      0.5,
		},
		{
			name:      "complex time",
			timeOfDay: &TimeOfDay{Hours: 2, Minutes: 30, Seconds: 45, Nanos: 123456789},
			want:      2*3600 + 30*60 + 45 + 0.123456789,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := tt.timeOfDay.TotalSeconds()
			assert.InDelta(t, tt.want, got, 0.000000001) // Allow small floating point differences
		})
	}
}

func TestNewTimeOfDayFromTime(t *testing.T) {
	tests := []struct {
		name     string
		timeVal  time.Time
		wantTime *TimeOfDay
	}{
		{
			name:     "midnight",
			timeVal:  time.Date(2023, 12, 25, 0, 0, 0, 0, time.UTC),
			wantTime: &TimeOfDay{Hours: 0, Minutes: 0, Seconds: 0, Nanos: 0},
		},
		{
			name:     "noon",
			timeVal:  time.Date(2023, 12, 25, 12, 0, 0, 0, time.UTC),
			wantTime: &TimeOfDay{Hours: 12, Minutes: 0, Seconds: 0, Nanos: 0},
		},
		{
			name:     "complex time",
			timeVal:  time.Date(2023, 12, 25, 15, 30, 45, 123456789, time.UTC),
			wantTime: &TimeOfDay{Hours: 15, Minutes: 30, Seconds: 45, Nanos: 123456789},
		},
		{
			name:     "end of day",
			timeVal:  time.Date(2023, 12, 25, 23, 59, 59, 999999999, time.UTC),
			wantTime: &TimeOfDay{Hours: 23, Minutes: 59, Seconds: 59, Nanos: 999999999},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := NewTimeOfDayFromTime(tt.timeVal)
			assert.Equal(t, tt.wantTime, got)
		})
	}
}
